From 07950edc26d548357d45ca61cb9b94353ea5b453 Mon Sep 17 00:00:00 2001
From: Shengzhe Zhao <a18689@motorola.com>
Date: Thu, 2 Jun 2016 15:50:32 -0700
Subject: [PATCH 903/959] IKSWM-43184 exfat: avoid setting 0 over buffer
 uniname

utf8s_to_utf16s could return -EINVAL(-22), then uniname[i] =
uniname[-22] = '\0' will corrupt other local variable.
uniname is a local variable having different address everytime
when call nls_cstring_to_uniname. so uniname[-22] could corrupt
some different addresses.
and also when filename reaches at maximum length, need to leave
last one space in buffer for '\0'

Change-Id: I932b6d246c1e6552f5e6331e04dd63ad97d058b6
Signed-off-by: Shengzhe Zhao <a18689@motorola.com>
Reviewed-on: http://gerrit.mot.com/863277
SLTApproved: Slta Waiver <sltawvr@motorola.com>
SME-Granted: SME Approvals Granted
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Zhi-Ming Yuan <a14194@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
Reviewed-on: http://gerrit.mot.com/864622
---
 fs/exfat/exfat_nls.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/fs/exfat/exfat_nls.c b/fs/exfat/exfat_nls.c
index 3cd559f7148..dedc212f6c9 100644
--- a/fs/exfat/exfat_nls.c
+++ b/fs/exfat/exfat_nls.c
@@ -296,6 +296,9 @@ void nls_uniname_to_cstring(struct super_block *sb, u8 *p_cstring, UNI_NAME_T *p
 
 	if (nls == NULL) {
 		len = utf16s_to_utf8s(uniname, MAX_NAME_LENGTH, UTF16_HOST_ENDIAN, p_cstring, MAX_NAME_LENGTH);
+		if (len >= MAX_NAME_LENGTH * MAX_CHARSET_SIZE)
+			len--;
+
 		p_cstring[len] = 0;
 		return;
 	}
@@ -360,7 +363,11 @@ void nls_cstring_to_uniname(struct super_block *sb, UNI_NAME_T *p_uniname, u8 *p
 #endif
 		for (j = 0; j < i; j++)
 			SET16_A(upname + j * 2, nls_upper(sb, uniname[j]));
-		uniname[i] = '\0';
+
+		if (i >= 0 && i < MAX_NAME_LENGTH)
+			uniname[i] = '\0';
+		else
+			lossy = TRUE;
 	}
 	else {
 		i = j = 0;
-- 
2.11.0

